diff --git a/themes/agnoster.zsh-theme b/themes/agnoster.zsh-theme
index 6345237..6ce57af 100644
--- a/themes/agnoster.zsh-theme
+++ b/themes/agnoster.zsh-theme
@@ -79,7 +79,7 @@ esac
 
 # VirtualEnv colors
 : ${AGNOSTER_VENV_FG:=black}
-: ${AGNOSTER_VENV_BG:=blue}
+: ${AGNOSTER_VENV_BG:=magenta}
 
 # AWS Profile colors
 : ${AGNOSTER_AWS_PROD_FG:=yellow}
@@ -230,7 +230,7 @@ prompt_git() {
     zstyle ':vcs_info:*' get-revision true
     zstyle ':vcs_info:*' check-for-changes true
     zstyle ':vcs_info:*' stagedstr '✚'
-    zstyle ':vcs_info:*' unstagedstr '±'
+    zstyle ':vcs_info:*' unstagedstr '●'
     zstyle ':vcs_info:*' formats ' %u%c'
     zstyle ':vcs_info:*' actionformats ' %u%c'
     vcs_info
@@ -357,6 +357,34 @@ prompt_terraform() {
   local terraform_info=$(tf_prompt_info)
   [[ -z "$terraform_info" ]] && return
   prompt_segment magenta yellow "TF: $terraform_info"
+
+}
+
+prompt_kubernetes_icon() {
+  KUBERNETES_SYMBOL=$'\xE2\x8E\x88'
+  KUBERNETES_CONTEXT=$1
+
+  case "$KUBERNETES_CONTEXT" in
+    *prod*) prompt_segment black red "$KUBERNETES_SYMBOL";;
+    *staging*) prompt_segment black yellow "$KUBERNETES_SYMBOL";;
+    *) prompt_segment black green "$KUBERNETES_SYMBOL";;
+  esac
+
+}
+
+prompt_kubecontext() {
+  KUBERNETES_BINARY="${KUBERNETES_BINARY:-kubectl}"
+  [[ -z "$KUBECONFIG" && -z "$(${KUBERNETES_BINARY} config current-context 2>/dev/null)" ]] && return
+
+  KUBERNETES_CONTEXT="$(${KUBERNETES_BINARY} config current-context 2> /dev/null)"
+  KUBERNETES_CONTEXT="${KUBERNETES_CONTEXT:-N/A}"
+  KUBERNETES_NAMESPACE="$(${KUBERNETES_BINARY} config view --minify -o jsonpath={..namespace} 2>/dev/null)"
+  KUBERNETES_NAMESPACE="${KUBERNETES_NAMESPACE:-default}"
+  KUBERNETES_NAMESPACE="${KUBERNETES_NAMESPACE:+:$KUBERNETES_NAMESPACE}"
+
+  prompt_kubernetes_icon "$KUBERNETES_CONTEXT"
+  prompt_segment $CURRENT_BG default "$KUBERNETES_CONTEXT$KUBERNETES_NAMESPACE"
+
 }
 
 ## Main prompt
@@ -366,6 +394,7 @@ build_prompt() {
   prompt_virtualenv
   prompt_aws
   prompt_terraform
+  prompt_kubecontext
   prompt_context
   prompt_dir
   prompt_git
